app/controllers/base_controller.ts

```diff
@@ -128,6 +146,83 @@ export default abstract class BaseController<
    */
   protected readonly singletonId?: number | string;
 
+  /**
+   * Return the permission slug required for the given action, or null/undefined if none is required.
+   * Override this in derived controllers to impose custom requirements per action.
+   * Defaults:
+   *  - store -> "create"
+   *  - update -> "update"
+   *  - destroy -> "destroy"
+   *  - relationIndex -> none (public by default)
+   *  - oneToManyRelationStore -> "create"
+   *  - manyToManyRelationAttach -> "update"
+   *  - manyToManyRelationDetach -> "update"
+   *  - index/show -> none (public by default)
+   */
+  // eslint-disable-next-line @typescript-eslint/no-unused-vars
+  protected requiredPermissionFor(
+    action: ControllerAction,
+  ): string | null | undefined {
+    switch (action) {
+      case "store":
+        return "create";
+      case "update":
+        return "update";
+      case "destroy":
+        return "destroy";
+      case "oneToManyRelationStore":
+        return "create";
+      case "manyToManyRelationAttach":
+      case "manyToManyRelationDetach":
+        return "update";
+      case "index":
+      case "show":
+      case "relationIndex":
+      default:
+        return null;
+    }
+  }
+
+  /**
+   * Authenticate and check permission for the given action.
+   * If no permission is required, this is a no-op and does not force authentication.
+   * Throws ForbiddenException on failure with minimal info.
+   */
+  protected async authenticate(
+    http: HttpContext,
+    action: ControllerAction,
+  ): Promise<void> {
+    const slug = this.requiredPermissionFor(action);
+    if (!slug) return; // public endpoint by default
+
+    if (!http.auth.isAuthenticated) {
+      await http.auth.authenticate();
+    }
+    // Superuser bypass: solvro_admin has access to everything
+    const isSolvroAdmin = await (http.auth.user as any)?.hasRole?.(
+      "solvro_admin",
+    );
+    if (isSolvroAdmin) return;
+    const has = await http.auth.user?.hasPermission(slug);
+    if (!has) {
+      throw new ForbiddenException();
+    }
+  }
+
+  /**
+   * Optional per-record authorization hook. Override to enforce row-level checks.
+   * Called after the record is fetched and before any data is returned or mutated.
+   */
+  protected async authorizeRecord(
+    http: HttpContext,
+    action: ControllerAction,
+    record: InstanceType<T>,
+  ): Promise<void> {
+    void http;
+    void action;
+    void record;
+  }
```

Comment:
authorization should ideally be done before interacting with the db, so it should ideally take an id, not a record